### ISSUE TITLE
Geolocation Datatypes

### ISSUE THREAD

**Issue by schickling** [prisma/prisma1#28]
* `lat` and `lng` float fields
* radius-based and region-based filtering of locations

Here are some more relevant/related links:

- [Postgis Prisma feature request](https://github.com/prisma/prisma/issues/1760)
- [Aurora spacial indexing](https://aws.amazon.com/blogs/aws/amazon-aurora-update-spatial-indexing-and-zero-downtime-patching/)
- [knex-postgis extension](https://github.com/jfgodoy/knex-postgis)
- [TypeORM feature request](https://github.com/typeorm/typeorm/issues/3296)
- [GeoAlchemy 2](https://geoalchemy-2.readthedocs.io/en/latest/)

**Comment by schickling** [comment]
Closed as dupe of #22

**Comment by marktani** [comment]
Reopening because #22 had a different scope.

**Comment by larryranches** [comment]
Any update on the roadmap to support the geolocation data type?

**Comment by marktani** [comment]
we're currently preparing an update to our roadmap :slightly_smiling_face: Could you join our [Slack](https://slack.graph.cool) and reach out to me there? My nickname ist @nilan.

**Comment by michaelpumo** [comment]
I would love to see this kind of ordering available baked-in with GraphCool.

I'm building an app for a well-known property company and it would be great to be able to list their property locations in an order to a given longitude and latitude. That way, users can find properties near them using HTML5 Geolocation API.

Apologies I don't have much to add in terms of how to achieve this, but it would be an amazing feature.

Thanks!

**Comment by angineering** [comment]
Hi, is there any update on the prioritisation of this feature? It would be really useful to have.

**Comment by anaibol** [comment]
+1

**Comment by movibe** [comment]
+1

**Comment by anaibol** [comment]
+1

**Comment by cckrishna** [comment]
+1
I am also looking for this capability in GraphCool. I have few cycles and can help with implementation and contribute back.  Let me know the next steps.

**Comment by ciocan** [comment]
+1

**Comment by saferthanhouses** [comment]
+1

**Comment by parweb** [comment]
+1

**Comment by AwoyoToyin** [comment]
+1

**Comment by mcmar** [comment]
+1

**Comment by ebouJ** [comment]
+1

**Comment by jhalborg** [comment]
+1, subscribing, indirect ping :-) Would love this feature!

An updated ETA/prioritization, or perhaps a guide on how to implement it without having to change too much when this is implemented would be great!

**Comment by hakimelek** [comment]
+1

**Comment by jtsubramanya** [comment]
we have couple of use case
a)  where we have to get all the sensors for a polygon passed as a input parameter to prisma.
ie,
1. assume i have stored sensors in a location like a city with lat long
2. instead of getting all the lights in a city i would create a polygon for a smaller area and want to get all the sensors which would fall inside that polygon.

b) we will store polygon in prisma for a location like city, now pass a geo cordinate of a sensor and get the location in which it falls in.

Basically looking for something what is there in mongoDB
https://docs.mongodb.com/manual/reference/operator/query/polygon/   
is there any way to achieve this in prisma?

**Comment by 4ware** [comment]
Yeah +1!!!! I really need geolocation especially queries for objects near a specific location and all objects within a radius...

**Comment by FluorescentHallucinogen** [comment]
It seems that Prisma team really intends to implement this. The only question: when?

![Geo Queries](https://user-images.githubusercontent.com/7892779/42722426-79f4e2d4-8754-11e8-88da-d4f044738bf5.png)

Source: https://www.prisma.io/features/databases/

**Comment by terijyu** [comment]
Late to the train..... but big +1!

**Comment by maxnucci** [comment]
🙌+1

**Comment by Findiglay** [comment]
+1

**Comment by kratam** [comment]
+1

**Comment by anaibol** [comment]
+1

**Comment by jsbrain** [comment]
+1

**Comment by arabscimitar** [comment]
any updates on this

**Comment by ian** [comment]
+1

**Comment by n0ne** [comment]
+1

**Comment by jimmyjamieson** [comment]
+1

**Comment by mrimran** [comment]
Any updates so far? Its now last month of 2018 :)

**Comment by priley86** [comment]
this would be a much appreciated feature (it's kind of a roadblock for many apps). The [workaround](https://stackoverflow.com/questions/51010009/does-prisma-supports-geospatial-queries) described on SO is far from ideal.

**Comment by dna1986** [comment]
+1

**Comment by BePasquet** [comment]
+1

**Comment by bannouheol** [comment]
+1

**Comment by yandongxu** [comment]
+1 waiting for this very useful feature

**Comment by lejager** [comment]
+1

**Comment by branislaav** [comment]
+1

**Comment by 213374you** [comment]
+1 any updates when this will be supported? At least for PostGIS

**Comment by clansofts** [comment]
Waiting for this feature..

**Comment by philipfeldmann** [comment]
+1

**Comment by gsancho** [comment]
+1

**Comment by vitalicwow** [comment]
This is preventing me from using it in my project

**Comment by 5an1ty** [comment]
Yes please, blocking for me as well!

**Comment by vidrepar** [comment]
+1

**Comment by erikostling** [comment]
+1

**Comment by jacob-ebey** [comment]
Are you going to give the community any direction on where your're going with this? Or are we to assume that we should move on?

**Comment by AmirBraham** [comment]
+1

**Comment by AmirBraham** [comment]
Has anyone found an alternative way to filter locations within a specific radius ?

**Comment by jacob-ebey** [comment]
@AmirBraham, until we hear a peep from someone on the Prisma team, you can do something similar to:
```graphql
input LocationInput {
  distance: Int
  lat: Float!
  lon: Float!
}

type Query {
  stores(location: LocationInput!): [Store!]!
}
```
```typescript
// constant somewhere
const EARTH_RADIUS_KM = 6371;

// stores query resolver
function stores(_: any, { location }, { prisma }) {
  const distance = location.distance || 10;

  const maxLat = location.lat + rad2Deg(distance / EARTH_RADIUS_KM);
  const minLat = location.lat - rad2Deg(distance / EARTH_RADIUS_KM);

  const maxLon = location.lon + rad2Deg(distance / EARTH_RADIUS_KM / Math.cos(deg2Rad(location.lat)));
  const minLon = location.lon - rad2Deg(distance / EARTH_RADIUS_KM / Math.cos(deg2Rad(location.lat)));

  return prisma.stores({
    where: {
      location: {
        AND: [
          {
            lat_gte: minLat,
            lat_lte: maxLat,
            lon_gte: minLon,
            lon_lte: maxLon
          }
        ]
      }
    }
  });
}
```

**Comment by AmirBraham** [comment]
@jacob-ebey  Thanks ! works like a charm ! 😄

**Comment by jacob-ebey** [comment]
@AmirBraham, no problem. It's a shame the Prisma team remains so evasive. (2017 seems to be the last we've heard from them on this)

**Comment by jacob-ebey** [comment]
lolol, this is not a stale issue. The Prisma team needs to respond.

**Comment by jacob-ebey** [comment]
Bump?

Any response would be better than nothing guys....

A simple "We are not going to support this feature" would suffice and I'll roll up the above solution into a shareable package for the community who has been asking for this feature for years.

Does this team even exist anymore?

**Comment by esistgut** [comment]
I guess most of the attention is now on https://github.com/prisma/prisma2

**Comment by jacob-ebey** [comment]
Also, your homepage is broken and 404ing on your js bundles....

**Comment by ontwikkelfabriek** [comment]
PostGIS support asp please

**Comment by rilexus** [comment]
+1

**Comment by ishmaelthedestroyer** [comment]
+1

**Comment by abhishekover9000** [comment]
+1000

**Comment by abhishekover9000** [comment]
Very pertinent issue to using Prisma as a complete ORM solution. An Elasticsearch integration would be awesome too!

> @AmirBraham, until we hear a peep from someone on the Prisma team, you can do something similar to:
> 
> Prisma schema
> 
> ```graphql
> type Location {
>   lat: Float!
>   lon: Float!
> }
> 
> type Store {
>   id: ID! @unique
>   label: String!
>   location: Location!
> }
> ```
> 
> Application schema
> 
> ```graphql
> input LocationInput {
>   distance: Float
>   lat: Float!
>   lon: Float!
> }
> 
> type Query {
>   stores(location: LocationInput!): [Store!]!
> }
> ```
> 
> Application resolver
> 
> ```ts
> // constant somewhere
> const EARTH_RADIUS_KM = 6371;
> 
> // stores query resolver
> function stores(_, { location }, { prisma }) {
>   const distance = location.distance || 10;
> 
>   const maxLat = location.lat + rad2Deg(distance / EARTH_RADIUS_KM);
>   const minLat = location.lat - rad2Deg(distance / EARTH_RADIUS_KM);
> 
>   const maxLon = location.lon + rad2Deg(distance / EARTH_RADIUS_KM / Math.cos(deg2Rad(location.lat)));
>   const minLon = location.lon - rad2Deg(distance / EARTH_RADIUS_KM / Math.cos(deg2Rad(location.lat)));
> 
>   return prisma.stores({
>     where: {
>       location: {
>         AND: [
>           {
>             lat_gte: minLat,
>             lat_lte: maxLat,
>             lon_gte: minLon,
>             lon_lte: maxLon
>           }
>         ]
>       }
>     }
>   });
> }
> ```

That resolver could be the basis of a prisma geosearch API - @jacob-ebey  since you came up with the original logic, do you want to build this into a PR?

**Comment by jacob-ebey** [comment]
@abhishekover9000 You can go ahead. I'm not really using Prisma anymore due to the lack or response on issues like this. That solution was pulled straight out of a google search, so whatever "C" "StackOverflow" question that answeres something along the lines of "Find lat lon within radius" should be credited hahaha

**Comment by vamshi9666** [comment]
This needs to exist. Any basic application needs this. Whats current status and is there any workarounds ? Only that one comment above?

**Comment by yodaheis** [comment]
True, planned migrating from Prisma now. Lots of issues like this have no updates altogether. On the other hand, newer ways/ products keep on popping up within Prisma ecosystem.
I guess Prisma team is trying to aim for too many things too fast, while not being production-ready choice for any one.

**Comment by carstenbaumhoegger** [comment]
+1

**Comment by cclaflin89** [comment]
+1

**Comment by jacob-ebey** [comment]
This is dead in the water (prove me wrong Prisma).

**Comment by hesa2020** [comment]
+1 👍

**Comment by ontwikkelfabriek** [comment]
+1

**Comment by AbanoubEMhanna** [comment]
+1

**Comment by meseven** [comment]
+1

**Comment by murbanowicz** [comment]
Could any maintainer please give the statemt if this is going to be implemented or not?

**Comment by yjarid** [comment]
+1

**Comment by EliasTouil** [comment]
+1

**Comment by piemonte** [comment]
+1

**Comment by ontwikkelfabriek** [comment]
almost 4 year and nothing. It's a joke. They don't care and don't read this. It's like shitting on your customers.

**Comment by ontwikkelfabriek** [comment]
+11

**Comment by dariotrombello** [comment]
+1

**Comment by geomin76** [comment]
+1

**Comment by sgup** [comment]
+1

**Comment by vdesdoigts** [comment]
Hello, has anyone found a solution? Thanks

**Comment by troykelly** [comment]
It feels like this issue maybe needs to get reopened until it's resolved?

**Comment by kaloyanBozhkov** [comment]
+1

**Comment by ngrhadi** [comment]
[https://www.prisma.io/docs/reference/database-reference/database-features#constraints](url)

When I see this, doc says using `@default` to define shapefile data, but in link doc about _shema data model default_ nothing introduce something like `@default.geom` / `@default.geometry` , any wrong with me?
[https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-a-default-value](url)

**Comment by jaimefps** [comment]
+1

**Comment by AllounSalah** [comment]
+1

**Comment by FeMaffezzolli** [comment]
+1

**Comment by dreglad** [comment]
Lack of this feature made me switched to https://hasura.io/

**Comment by pschaub** [comment]
This issue is part of Prisma v1. If you are using Prisma v2 please have a look into https://github.com/prisma/prisma/issues/2789
